be/src/exec/pipeline/sink/connector_sink_operator.cpp

```diff
@@ -0,0 +1,154 @@
+            return false;
+        }
+
+        if (auto st = _add_chunk_future_queue.front().get(); !st.ok()) {
+            LOG(WARNING) << "cancel fragment: " << st;
+            _fragment_context->cancel(st);
+        }
+        _add_chunk_future_queue.pop();
+    }
+
+    while (!_commit_file_future_queue.empty()) {
+        // unfinished if any commit_file_future future is not ready
+        if (!is_ready(_commit_file_future_queue.front())) {
+            return false;
+        }
+
+        auto result = _commit_file_future_queue.front().get();
+        _commit_file_future_queue.pop();
+
+        if (auto st = result.io_status; st.ok()) {
+            // invoke callback if file commit succeed
+            _connector_chunk_sink->callback_on_success()(result);
+        } else {
+            LOG(WARNING) << "cancel fragment: " << st;
+            _fragment_context->cancel(st);
+        }
+        _rollback_actions.push(std::move(result.rollback_action));
+    }
+
+    DCHECK(_add_chunk_future_queue.empty());
+    DCHECK(_commit_file_future_queue.empty());
+    return true;
+}
+
+Status ConnectorSinkOperator::set_finishing(RuntimeState* state) {
+    _no_more_input = true;
+    auto future = _connector_chunk_sink->finish();
+    _enqueue_futures(std::move(future));
+    return Status::OK();
+}
+
+bool ConnectorSinkOperator::pending_finish() const {
+    return !is_finished();
+}
+
+Status ConnectorSinkOperator::set_cancelled(RuntimeState* state) {
+    _is_cancelled = true;
+    return Status::OK();
+}
+
+StatusOr<ChunkPtr> ConnectorSinkOperator::pull_chunk(RuntimeState* state) {
+    CHECK(false) << "ConnectorSinkOperator::pull_chunk";
+    __builtin_unreachable();
+}
+
+Status ConnectorSinkOperator::push_chunk(RuntimeState* state, const ChunkPtr& chunk) {
+    ASSIGN_OR_RETURN(auto future, _connector_chunk_sink->add(chunk));
+    _enqueue_futures(std::move(future));
+    return Status::OK();
+}
+
+void ConnectorSinkOperator::_enqueue_futures(connector::ConnectorChunkSink::Futures future) {
+    for (auto& f : future.add_chunk_future) {
+        _add_chunk_future_queue.push(std::move(f));
+    }
+    for (auto& f : future.commit_file_future) {
+        _commit_file_future_queue.push(std::move(f));
+    }
+}
+
+ConnectorSinkOperatorFactory::ConnectorSinkOperatorFactory(
+        int32_t id, std::unique_ptr<connector::ConnectorChunkSinkProvider> data_sink_provider,
+        std::shared_ptr<connector::ConnectorChunkSinkContext> sink_context, FragmentContext* fragment_context)
+        : OperatorFactory(id, "connector sink operator", Operator::s_pseudo_plan_node_id_for_final_sink),
+          _data_sink_provider(std::move(data_sink_provider)),
+          _sink_context(sink_context),
+          _fragment_context(fragment_context) {}
+
+OperatorPtr ConnectorSinkOperatorFactory::create(int32_t degree_of_parallelism, int32_t driver_sequence) {
+    auto chunk_sink = _data_sink_provider->create_chunk_sink(_sink_context, driver_sequence);
```

Comment:
create_chunk_sink is called in fe->be brpc service handler threads, we must guarantee that this function would not  hold the thread too much time.